File: tools/dbk_extract.py
#!/usr/bin/env python3
"""
Extract Sharp Zaurus .dbk whole-NAND images into main-area binaries.

Observed format for systc760.dbk:
- 16-byte global header
- 8192 block records for a 128MiB NAND (16KiB eraseblock)
- each record is 0x4210 bytes: 0x10 per-block header + 0x4200 page payload
- payload contains 32 pages of (512B main + 16B OOB)

This tool reconstructs:
- full_main.bin (128MiB main area)
- smf_main.bin  (first 0x700000 bytes)
- root_main.bin (next 0x3500000 bytes)
- home_main.bin (remaining 0x4400000 bytes)
- optional per-block headers dump for forensics
"""

from __future__ import annotations

import argparse
import os
import struct
from pathlib import Path

GLOBAL_HDR = 0x10
BLOCK_REC = 0x4210
BLOCK_HDR = 0x10
BLOCK_PAYLOAD = 0x4200
PAGE_MAIN = 512
PAGE_OOB = 16
PAGES_PER_BLOCK = 32

NAND_BLOCKS = 8192
FULL_MAIN_SIZE = 0x8000000  # 128MiB
SMF_SIZE = 0x700000
ROOT_SIZE = 0x3500000
HOME_SIZE = 0x4400000


def iter_blocks(buf: bytes):
    off = GLOBAL_HDR
    for idx in range(NAND_BLOCKS):
        rec = buf[off : off + BLOCK_REC]
        if len(rec) != BLOCK_REC:
            raise ValueError(f"short record at block {idx}")
        hdr = rec[:BLOCK_HDR]
        payload = rec[BLOCK_HDR:]
        yield idx, hdr, payload
        off += BLOCK_REC


def payload_to_main(payload: bytes) -> bytes:
    if len(payload) != BLOCK_PAYLOAD:
        raise ValueError("invalid payload size")
    out = bytearray(PAGES_PER_BLOCK * PAGE_MAIN)
    src = 0
    dst = 0
    for _ in range(PAGES_PER_BLOCK):
        out[dst : dst + PAGE_MAIN] = payload[src : src + PAGE_MAIN]
        src += PAGE_MAIN + PAGE_OOB
        dst += PAGE_MAIN
    return bytes(out)


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("dbk", help="path to .dbk file")
    ap.add_argument("--out-dir", default="flash/.cache/dbk", help="output directory")
    args = ap.parse_args()

    dbk_path = Path(args.dbk)
    out_dir = Path(args.out_dir)
    out_dir.mkdir(parents=True, exist_ok=True)

    data = dbk_path.read_bytes()
    expected = GLOBAL_HDR + NAND_BLOCKS * BLOCK_REC
    if len(data) != expected:
        raise SystemExit(
            f"unexpected dbk size: got {len(data)}, expected {expected}"
        )

    # Basic marker sanity.
    if not data.startswith(b"[I]JFFS2TOP:"):
        raise SystemExit("unexpected global header magic")

    full_main = bytearray(FULL_MAIN_SIZE)
    headers_path = out_dir / "block_headers.bin"
    with headers_path.open("wb") as hfp:
        for idx, hdr, payload in iter_blocks(data):
            hfp.write(hdr)
            block_main = payload_to_main(payload)
            start = idx * (PAGES_PER_BLOCK * PAGE_MAIN)
            full_main[start : start + len(block_main)] = block_main

    full_main_path = out_dir / "full_main.bin"
    full_main_path.write_bytes(full_main)

    smf = full_main[:SMF_SIZE]
    root = full_main[SMF_SIZE : SMF_SIZE + ROOT_SIZE]
    home = full_main[SMF_SIZE + ROOT_SIZE : SMF_SIZE + ROOT_SIZE + HOME_SIZE]

    (out_dir / "smf_main.bin").write_bytes(smf)
    (out_dir / "root_main.bin").write_bytes(root)
    (out_dir / "home_main.bin").write_bytes(home)

    print(f"dbk: {dbk_path}")
    print(f"size: {len(data)} bytes")
    print(f"full_main: {full_main_path} ({len(full_main)} bytes)")
    print(f"smf_main: {(out_dir / 'smf_main.bin')} ({len(smf)} bytes)")
    print(f"root_main: {(out_dir / 'root_main.bin')} ({len(root)} bytes)")
    print(f"home_main: {(out_dir / 'home_main.bin')} ({len(home)} bytes)")
    print(f"headers: {headers_path} ({headers_path.stat().st_size} bytes)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/scripts/dbk-extract.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');

const GLOBAL_HDR = 0x10;
const BLOCK_REC = 0x4210;
const BLOCK_HDR = 0x10;
const BLOCK_PAYLOAD = 0x4200;
const PAGE_MAIN = 512;
const PAGES_PER_BLOCK = 32;

const NAND_BLOCKS = 8192;
const FULL_MAIN_SIZE = 0x8000000;
const SMF_SIZE = 0x700000;
const ROOT_SIZE = 0x3500000;
const HOME_SIZE = 0x4400000;

function parseArgs(argv) {
    const args = { outDir: 'flash/.cache/dbk' };
    const positional = [];
    for (let i = 0; i < argv.length; i++) {
        if (argv[i] === '--out-dir') {
            args.outDir = argv[++i];
        } else {
            positional.push(argv[i]);
        }
    }
    if (positional.length !== 1) {
        console.error('usage: dbk-extract.js <path-to.dbk> [--out-dir DIR]');
        process.exit(1);
    }
    args.dbk = positional[0];
    return args;
}

function payloadToMain(payload) {
    if (payload.length !== BLOCK_PAYLOAD) {
        throw new Error('invalid payload size');
    }
    return payload.subarray(0, PAGES_PER_BLOCK * PAGE_MAIN);
}

function main() {
    const args = parseArgs(process.argv.slice(2));
    fs.mkdirSync(args.outDir, { recursive: true });

    const data = fs.readFileSync(args.dbk);
    const expected = GLOBAL_HDR + NAND_BLOCKS * BLOCK_REC;
    if (data.length !== expected) {
        console.error(`unexpected dbk size: got ${data.length}, expected ${expected}`);
        process.exit(1);
    }

    if (data.toString('ascii', 0, 12) !== '[I]JFFS2TOP:') {
        console.error('unexpected global header magic');
        process.exit(1);
    }

    const fullMain = Buffer.alloc(FULL_MAIN_SIZE);
    const headersPath = path.join(args.outDir, 'block_headers.bin');
    const headersFd = fs.openSync(headersPath, 'w');

    let off = GLOBAL_HDR;
    for (let idx = 0; idx < NAND_BLOCKS; idx++) {
        const rec = data.subarray(off, off + BLOCK_REC);
        if (rec.length !== BLOCK_REC) {
            throw new Error(`short record at block ${idx}`);
        }
        const hdr = rec.subarray(0, BLOCK_HDR);
        const payload = rec.subarray(BLOCK_HDR);
        fs.writeSync(headersFd, hdr);

        const blockMain = payloadToMain(payload);
        const start = idx * (PAGES_PER_BLOCK * PAGE_MAIN);
        blockMain.copy(fullMain, start);

        off += BLOCK_REC;
    }
    fs.closeSync(headersFd);

    const fullMainPath = path.join(args.outDir, 'full_main.bin');
    fs.writeFileSync(fullMainPath, fullMain);

    const smf = fullMain.subarray(0, SMF_SIZE);
    const root = fullMain.subarray(SMF_SIZE, SMF_SIZE + ROOT_SIZE);
    const home = fullMain.subarray(SMF_SIZE + ROOT_SIZE, SMF_SIZE + ROOT_SIZE + HOME_SIZE);

    const smfPath = path.join(args.outDir, 'smf_main.bin');
    const rootPath = path.join(args.outDir, 'root_main.bin');
    const homePath = path.join(args.outDir, 'home_main.bin');
    fs.writeFileSync(smfPath, smf);
    fs.writeFileSync(rootPath, root);
    fs.writeFileSync(homePath, home);

    console.log(`dbk: ${args.dbk}`);
    console.log(`size: ${data.length} bytes`);
    console.log(`full_main: ${fullMainPath} (${fullMain.length} bytes)`);
    console.log(`smf_main: ${smfPath} (${smf.length} bytes)`);
    console.log(`root_main: ${rootPath} (${root.length} bytes)`);
    console.log(`home_main: ${homePath} (${home.length} bytes)`);
    console.log(`headers: ${headersPath} (${fs.statSync(headersPath).size} bytes)`);
}

main();
